[CURL] Access CURL object via `CFileItem::GetURL` and `CFileItem::GetDynURL`.

damienruscoe opened the pull request:
## Description
Following #26873 `CFileItem::GetURL()` and `CFileItem::GetDynURL()` methods have been updated to be very cheap to call.

As it is now possible to cheaply acquire a `CURL` object from a `CFileItem` object it will be more performant to add utility methods onto the CURL object. These methods will be better adapted than the `URIUtils::*` functions.

## Motivation and context

## How has this been tested?
Manually: A TV show library scan.
Added a comprehensive range of tests which verify URIUtils::*() and CURL::*() consistency

## What is the effect on users?

## Screenshots (if appropriate):

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

damienruscoe commented:
Dependent upon #27087

damienruscoe commented:
Jenkins build this please

damienruscoe commented:
Jenkins build this please

I lost the failing build logs

damienruscoe commented:
@garbear Unsure if you are the correct person for this issue, appologies if I got the wrong person.

The Win64 build is driving me crazy for this branch.

I am trying to add `/bigobj` to the `kodi-test` when building with MSVC but all my attempts still lead to the same error message:

```
18:45:00 C:\jenkws\workspace\WIN-64\xbmc\utils\test\TestUrlParsing.cpp(1,1): error C1128: number of sections exceeded object file format limit: compile with /bigobj 
```

This is coming from the Jenkins build which I have little control over. Is this a CMake cache issue on the CI machines? Can I please get some direction to a way in this could be overcome?

neo1973 commented:
This simply generates way to much code as is. On my machine with 16GB of memory I can't build this with clang and enabled sanitizers because the compiler runs out of memory. If I replace all the `std::string` in `TestURLParseDetailsData` with `std::string_view` it builds without issues. Maybe that helps with MSVC as well.

neo1973 commented:
Now it compiles, and it was a fun little experiment, but in practice this isn't viable this was. I just had to abort a release build of the unit test after GCC was chewing for over 20 min on TestUrlParsing.cpp alone.

The only realistic solution is to load the tests at runtime instead of compiling them in. I suggest to convert them to JSON. And then the number of tests should probably be reduced, that many are simply not maintainable. If we decide to change the behavior at some point nobody will edit >700 tests.

neo1973 commented on xbmc/FileItem.cpp:
```diff
@@ -336,15 +336,23 @@ CFileItem::CFileItem(const CURL& path, bool bIsFolder) : m_strPath(path.Get())
   SetFolder(bIsFolder);
   if (bIsFolder && !m_strPath.empty() && !IsFileFolder())
-    URIUtils::AddSlashAtEnd(m_strPath);
+  {
+    std::string folderPath = m_strPath;
+    URIUtils::AddSlashAtEnd(folderPath);
+    SetPath(folderPath);
+  }
```
IMO it would be OK to change this to
```cpp
{
    URIUtils::AddSlashAtEnd(m_strPath);
    m_urlPath.reset();
}
```

neo1973 commented on xbmc/URL.cpp:
```diff
@@ -803,3 +804,191 @@ void CURL::RemoveProtocolOption(const std::string &key)
   m_strProtocolOptions = m_protocolOptions.GetOptionsString(false);
 }
+
+bool CURL::HasExtension(const std::string& strExtensions) const
```
```suggestion
bool CURL::HasExtension(std::string_view strExtensions) const
```

damienruscoe replied:
Fixed

neo1973 commented on xbmc/URL.cpp:
```diff
@@ -803,3 +804,191 @@ void CURL::RemoveProtocolOption(const std::string &key)
+  if (pos == std::string::npos || m_strFileName[pos] != '.')
+    return false;
+
+  const std::string extensionLower = StringUtils::ToLower(m_strFileName.substr(pos));
+
+  const std::vector<std::string> extensionsLower =
+      StringUtils::Split(StringUtils::ToLower(strExtensions), '|');
+
+  for (const auto& ext : extensionsLower)
+  {
+    if (StringUtils::EndsWith(ext, extensionLower))
+      return true;
+  }
```
The function becomes a lot more efficient if a range view is used instead of our `StringUtils::Split`:
```cpp
{
  const size_t pos = m_strFileName.find_last_of("./\\");
  if (pos == std::string::npos || m_strFileName[pos] != '.')
    return false;

  const std::string_view extension(m_strFileName.begin() + pos, m_strFileName.end());

  auto extensions =
      std::ranges::split_view(strExtensions, '|') |
      std::views::transform([](const auto& r) { return std::string_view(r.begin(), r.end()); });

  return std::ranges::any_of(extensions, [&](std::string_view ext)
                             { return StringUtils::EqualsNoCase(ext, extension); });
}
```

damienruscoe replied:
I like it and can see where the efficency is coming from.
You could also lowercase before the any_of loop and do a regular comparison inside the loop.

This does not compile with my compiler and I am tired of fighting compilers for this branch. I will definitly revisit this later.

As an aside, I think ranges may be of benefit in `GetDirectory`. We currently iterate through the directory items multiple times and ranges may reduce this to a single iteration.

damienruscoe commented on xbmc/utils/test/TestUrlParsing.cpp:
```diff
@@ -0,0 +1,526 @@
+{
+};
+
+/*
+#include <fstream>
+void write(std::ostream& f, const std::string& key, const int& value)
+{
+	if (value != 0)
+	  f << "  ." << key << " = " << value << ",\n";
+}
```
Agreed. Was looking for a complete build before removing.

damienruscoe commented on CMakeLists.txt:
```diff
@@ -577,6 +577,10 @@ if(HOST_CAN_EXECUTE_TARGET AND ENABLE_TESTING)
+    target_compile_options(${APP_NAME_LC}-test PRIVATE "/bigobj")
+  endif()
+
   set_target_properties(PROPERTIES ENABLE_EXPORTS ON)
```
Not considering. I think this suggestion is wrong. If not, these are going to be removed anyway.

damienruscoe commented:
@neo1973 
These two code suggestions have been addressed in resolved issues elsewhere on this page.

damienruscoe committed:
Regenerated test cases. Added input to json files.

damienruscoe commented:
Claude is hallucinating with the `HasExtension` logic bug. It has been correctly identified that this is probably not an optimal implementation but this was the existing implementation. This PR is large enough and does not have the intent of changing and behaviour or fixing any perfomance issues.

Good to get feedback on this branch if there is any. Thanks all

damienruscoe committed:
[CURL] Access CURL object via `CFileItem::GetURL` and `CFileItem::GetDynURL`.

These methods have been updated to be very cheap to call.

As it is now possible to cheaply acquire a `CURL` object from a
`CFileItem` object it will be more performant to add utility methods
onto the CURL object. These methods will be better adapted than the
`URIUtils::*` functions.

Added tests for consistency between these equivalent operations to
guarentee they are safe to switch between freely.

Added JSON based tests.

garbear commented:
@claude please review

damienruscoe committed:
Add documentation

garbear commented:
@damienruscoe Ready to go?
